fe/fe-core/src/main/java/com/starrocks/sql/common/PartitionDiffer.java

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.common;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Range;
+import com.starrocks.catalog.Column;
+import com.starrocks.catalog.MaterializedView;
+import com.starrocks.catalog.PartitionInfo;
+import com.starrocks.catalog.PartitionKey;
+import com.starrocks.catalog.PrimitiveType;
+import com.starrocks.catalog.RangePartitionInfo;
+import com.starrocks.catalog.Type;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.NotImplementedException;
+import com.starrocks.common.util.RangeUtils;
+import com.starrocks.scheduler.TaskRun;
+import com.starrocks.scheduler.TaskRunContext;
+import org.apache.commons.collections4.ListUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+// TODO: refactor all related code into this class
+
+/**
+ * Compute the Materialized View partition mapping from base table
+ * e.g. the MV is PARTITION BY date_trunc('month', dt), and the base table is daily partition, the resulted mv
+ * should be monthly partition
+ */
+public class PartitionDiffer {
```

Comment:
how about change this name to PartitionContext or PartitionRequestContext and move all these util function into SyncPartitionUtils?

Reply:
Good suggestion, and that's my first idea.

After that I realize that, the PartitionDiffer is much complicated than a stateless `Util`, which needs to support various functions and data types, and it needs many context information for partition pruning.

So I prefer to centralize these code into a `PartitionDiffer`, which is responsible for calculate the MV partition info based on the table partition and expression. In this way it would be more extensible to support more partitions types like ListPartition, and more partition functions.